### ISSUE TITLE
[OpenSSL] Update version to upcoming 1.1.1 (Bugzilla Bug 1089)

### ISSUE THREAD

**Issue by tianocore-issues** [tianocore/edk2#7432]
This issue was created automatically with [bugzilla2github](https://github.com/makubacki/bugzilla2github/tree/bz2gh_tianocore)
# Bugzilla Bug 1089
Date: 2018-08-15T13:30:37+00:00
From: P J P &lt;<prasad>&gt;
To:   Xiaoyu Lu &lt;<xiaoyux.lu>&gt;
CC:   jbuchert+tianocore, jian.j.wang, @lersek, qin.long, ting.ye, xiaoyux.lu, yonghong.zhu

Depends on:   #1845
Blocker for:  #1781, #1821, #1861
See also:     https://bugzilla.tianocore.org/show_bug.cgi?id=1926
Last updated: 2019-06-20T01:46:50+00:00

**Comment by tianocore-issues** [comment]
## Comment 4587
Date: 2018-08-15 13:30:37 +0000
From: P J P &lt;<prasad>&gt;

- Industry Specification: ---
- Release Observed: EDK II Master
- Releases to Fix: EDK II Master
- Target OS: ---
- Bugzilla Assignee(s): Xiaoyu Lu &lt;<xiaoyux.lu>&gt;

OpenSSL 1.1.1 Beta 6 (pre release 8) is available from https://www.openssl.org/
Upcoming 1.1.1 release will be included in Fedora soon.

This bug is to ensure that edk2 builds &amp; works with the new OpenSSL release.

**Comment by tianocore-issues** [comment]
## Comment 4611
Date: 2018-08-16 20:10:54 +0000
From: Yonghong Zhu &lt;<yonghong.zhu>&gt;

Bug scrub: Assign to module owner

**Comment by tianocore-issues** [comment]
## Comment 4989
Date: 2018-09-17 06:22:54 +0000
From: @lersek

OpenSSL 1.1.1 (LTS) was released on 11-Sep-2018.

**Comment by tianocore-issues** [comment]
## Comment 4991
Date: 2018-09-17 12:49:43 +0000
From: Qin Long &lt;<qin.long>&gt;

Thanks for reminder. 
Working on the 1.1.1 enabling &amp; evaluations ...

**Comment by tianocore-issues** [comment]
## Comment 5833
Date: 2018-12-12 02:14:01 +0000
From: Wang, Jian J &lt;<jian.j.wang>&gt;

Take the ownership

**Comment by tianocore-issues** [comment]
## Comment 6642
Date: 2019-03-07 12:07:05 +0000
From: @lersek

Hi Jian,

can you give me a time frame for advancing upstream edk2 to OpenSSL-1.1.1?

Does it perhaps require API level changes in the UEFI spec too (due to OpenSSL-1.1.1 implementing TLSv1.3)?

Thanks!
Laszlo

**Comment by tianocore-issues** [comment]
## Comment 6647
Date: 2019-03-08 01:53:24 +0000
From: Wang, Jian J &lt;<jian.j.wang>&gt;

(In reply to Laszlo Ersek from comment #5)

It will catch the next stable tag (Q2). Since current TlsLib can be built against openssl-1.1.1 without any change (we&apos;ve completed the initial porting for just build), we tend to keep TlsLib api as-is for backward compatibility but just update the implementation of it to meet Tls1.3 requirement. If we need to change existing api prototype for the sake of new feature of Tls1.3, normally a new library class will be introduced. This part is still under discussion. Any suggestion is welcomed. (Jian)

**Comment by tianocore-issues** [comment]
## Comment 6655
Date: 2019-03-08 04:07:41 +0000
From: @lersek

(In reply to Wang, Jian J from comment #6)
&gt; It will catch the next stable tag (Q2).

Thank you, this is helpful.

&gt; Since current TlsLib can be built against openssl-1.1.1 without any
&gt; change (we&apos;ve completed the initial porting for just build), we tend
&gt; to keep TlsLib api as-is for backward compatibility but just update
&gt; the implementation of it to meet Tls1.3 requirement.

That sounds great. My initial goal is not to utilize new features from
OpenSSL-1.1.1, but simply to build edk2 against OpenSSL-1.1.1.

Do you think it would be possible to post *just* an OpenSSL update, soon
after edk2-stable201903? Possibly with small TlsLib updates, as
required?

Again, my goal is simply to consume OpenSSL-1.1.1, through the current
library classes.

Thanks!

**Comment by tianocore-issues** [comment]
## Comment 6657
Date: 2019-03-08 07:31:09 +0000
From: Wang, Jian J &lt;<jian.j.wang>&gt;

(In reply to Laszlo Ersek from comment #7)
&gt; 
&gt; Do you think it would be possible to post *just* an OpenSSL update, soon
&gt; after edk2-stable201903? Possibly with small TlsLib updates, as
&gt; required?
&gt; 
&gt; Again, my goal is simply to consume OpenSSL-1.1.1, through the current
&gt; library classes.
&gt; 

I&apos;m afraid it cannot be *soon* because of the effort needed, especially the test effort. We just let the code pass build and we don&apos;t know whether it works or not. If you still want to try, I can post the work we&apos;ve done so far in my fork (there&apos;re reckless code just for build).

**Comment by tianocore-issues** [comment]
## Comment 6658
Date: 2019-03-08 08:29:09 +0000
From: @lersek

Hi Jian,

(In reply to Wang, Jian J from comment #8)
&gt; (In reply to Laszlo Ersek from comment #7)
&gt; &gt; 
&gt; &gt; Do you think it would be possible to post *just* an OpenSSL update, soon
&gt; &gt; after edk2-stable201903? Possibly with small TlsLib updates, as
&gt; &gt; required?
&gt; &gt; 
&gt; &gt; Again, my goal is simply to consume OpenSSL-1.1.1, through the current
&gt; &gt; library classes.
&gt; &gt; 
&gt; 
&gt; I&apos;m afraid it cannot be *soon* because of the effort needed, especially the
&gt; test effort.

Understood. It makes sense.

&gt; We just let the code pass build and we don&apos;t know whether it
&gt; works or not. If you still want to try, I can post the work we&apos;ve done so
&gt; far in my fork (there&apos;re reckless code just for build).

Thanks, my plate is full, I&apos;m not looking for more things to do. :) I now understand the situation, and I agree that regression testing is important. Let&apos;s re-raise this topic later, then.

Thank you!

**Comment by tianocore-issues** [comment]
## Comment 6789
Date: 2019-03-20 22:51:59 +0000
From: Yonghong Zhu &lt;<yonghong.zhu>&gt;

Per edk2-stable201905 planning to add the deadline info.

**Comment by tianocore-issues** [comment]
## Comment 6946
Date: 2019-03-29 02:31:35 +0000
From: Xiaoyu Lu &lt;<xiaoyux.lu>&gt;

Created attachment 283
git format patch files and test code

I made two patches for this.
0001-CryptoPkg-update-openssl-to-1.1.1b.patch
0002-CryptoPkg-Add-SM3-wrapper-function.patch

Cryptest.zip is in order to test these patches.

Attachment: [BZ-283-update_openssl_to_1_1_1b.zip](https://github.com/tianocore/user-attachments/raw/refs/heads/main/tianocore/edk2/BZ-283-update_openssl_to_1_1_1b.zip)

**Comment by tianocore-issues** [comment]
## Comment 6976
Date: 2019-04-02 05:24:50 +0000
From: @lersek

(In reply to xiaoyu from comment #11)
&gt; Created attachment 283 [details]
&gt; git format patch files and test code
&gt; 
&gt; I made two patches for this.
&gt; 0001-CryptoPkg-update-openssl-to-1.1.1b.patch
&gt; 0002-CryptoPkg-Add-SM3-wrapper-function.patch
&gt; 
&gt; Cryptest.zip is in order to test these patches.

(1) This BZ is currently assigned to Jian -- is that still correct? If not, please update the Assignee field.

(2) Reviewing patches as BZ attachments, and in BZ comments, is unwieldy. Please submit the patches to the edk2-devel mailing list, if you would like to receive comments at this point, even if you don&apos;t consider the patches &quot;final&quot; at this point -- just use the [RFC] subject-prefix, rather than [PATCH].

(3) I looked very-very briefly at patch#1. Please split it into multiple patches, one per module (BaseCryptLib, IntrinsicLib, OpensslLib, ...).

Preferably, the series should build at every step, against OpenSSL_1_1_0j too -- if that is possible at all --, and then the OpenSSL git submodule reference should be bumped to &quot;OpenSSL_1_1_1b&quot;, as the final step in the series.

(4) I think it would be useful to add the crypttest module to upstream edk2 as well. In my opinion, AppPkg could be a good home for it. Other reviewers may have better suggestions though (perhaps a separate repository even); I suggest discussing that on the list as well.

&quot;Cryptest.inf&quot; says at the top, in a comment: &quot;based on OpenSSL-1.0.2j&quot;. That implies that the test utility has existed since at least edk2 commit dab62c5ec8a8 (&quot;CryptoPkg/OpensslLib: Upgrade OpenSSL version to 1.0.2j&quot;, 2016-09-30). But this is the first time I hear about it.

**Comment by tianocore-issues** [comment]
## Comment 7015
Date: 2019-04-07 23:13:20 +0000
From: Wang, Jian J &lt;<jian.j.wang>&gt;

(In reply to Laszlo Ersek from comment #12)
(1) I&apos;ll transfer this tracker to Xiaoyu.
(2) Yes. Xiaoyu is new on the open source development. I&apos;ll guide him on how to send patch in mailling list.
(3) see (2)
(4) There&apos;s a edk2-test repo (https://github.com/tianocore/edk2-test) for this purpose. There&apos;s a team planning it but no idea how long it will take.

**Comment by tianocore-issues** [comment]
## Comment 7471
Date: 2019-05-09 08:53:45 +0000
From: @lersek

v1 series:
[edk2-devel] [PATCH 0/3] CryptoPkg: Upgrade OpenSSL to 1_1_1b
https://edk2.groups.io/g/devel/message/39750
Message-ID: &lt;1556525727-14875-1-git-send-email-xiaoyux.lu@intel.com&gt;

v2 series:
[edk2-devel] [PATCH v2 1/6] CryptoPkg/OpensslLib: Modify process_files.pl
                            for upgrading OpenSSL
https://edk2.groups.io/g/devel/message/40266
Message-ID: &lt;1557379429-7527-1-git-send-email-xiaoyux.lu@intel.com&gt;

**Comment by tianocore-issues** [comment]
## Comment 7583
Date: 2019-05-16 12:36:54 +0000
From: @lersek

v3 series:
[edk2-devel] [PATCH v3 0/6] CryptoPkg: Upgrade OpenSSL to 1.1.1b
http://mid.mail-archive.com/1557753912-30122-1-git-send-email-xiaoyux.lu@intel.com
https://edk2.groups.io/g/devel/message/40500

v4 series:
[edk2-devel] [PATCH v4 0/7] CryptoPkg: Upgrade OpenSSL to 1.1.1b
http://mid.mail-archive.com/1557993298-22205-1-git-send-email-xiaoyux.lu@intel.com
https://edk2.groups.io/g/devel/message/40761

**Comment by tianocore-issues** [comment]
## Comment 7860
Date: 2019-05-29 23:47:48 +0000
From: Xiaoyu Lu &lt;<xiaoyux.lu>&gt;

v5 series:
[PATCH v5 0/9] Crypto: Upgrade OpenSSL to 1.1.1b 
https://edk2.groups.io/g/devel/message/41587

Task BZ1781(https://bugzilla.tianocore.org/show_bug.cgi?id=1781) use SM3 algorithm from BaseCryptLib. But BaseCryptLib hasn&apos;t wrap SM3 algorithm from OpenSSL in v5 series.

So what should I do? 

(1): wrap OpenSSL SM3 algorithm to BaseCryptLib and send series v6.
(2): file a new BZ about wrap OpenSSL SM3 algorithm and send a patch separately.

I prefer (2).

What are your opinions?

**Comment by tianocore-issues** [comment]
## Comment 7861
Date: 2019-05-29 23:52:08 +0000
From: Wang, Jian J &lt;<jian.j.wang>&gt;

(In reply to Xiaoyu Lu from comment #16)
Please file a new BZ and separate patches for SM3. This one is just for openssl upgrading.

**Comment by tianocore-issues** [comment]
## Comment 7908
Date: 2019-06-03 07:41:26 +0000
From: @lersek

Based on &lt;http://mid.mail-archive.com/D827630B58408649ACB04F44C51000362590E91F@SHSMSX107.ccr.corp.intel.com&gt;, this issue has been fixed in commit range 371e7001e8d5..acfb90911840.

**Comment by tianocore-issues** [comment]
## Comment 7911
Date: 2019-06-03 13:04:24 +0000
From: @lersek

Fixup on top, for CLANG35:

f03859ea6c8f (&quot;CryptoPkg CLANG35: add -std=c99 to dodge OpenSSL C atomics issue&quot;, 2019-06-03)
